src/main/java/org/fcrepo/auth/webac/WebACRolesProvider.java

```diff
@@ -0,0 +1,299 @@
+        final List<URI> rdfTypes = resource.getTypes();
+
+        // Add the resource location and types of the ACL-bearing parent,
+        // if present and if different than the target resource.
+        effectiveAcl
+            .map(x -> x.getRight())
+            .filter(x -> !x.getPath().equals(resource.getPath()))
+            .ifPresent(x -> {
+                resourcePaths.add(FEDORA_INTERNAL_PREFIX + x.getPath());
+                rdfTypes.addAll(x.getTypes());
+            });
+
+        // Create a function to check acl:accessTo, scoped to the given resourcePaths
+        final Predicate<WebACAuthorization> checkAccessTo = accessTo.apply(resourcePaths);
+
+        // Create a function to check acl:accessToClass, scoped to the given rdf:type values;
+        // but transform the URIs to Strings first.
+        final Predicate<WebACAuthorization> checkAccessToClass =
+            accessToClass.apply(rdfTypes.stream().map(URI::toString).collect(Collectors.toList()));
+
+        // Read the effective Acl and return a list of acl:Authorization statements
+        final List<WebACAuthorization> authorizations = effectiveAcl
+                .map(uncheck(x -> getAuthorizations(x.getLeft().toString())))
+                .orElse(new ArrayList<>());
+
+        // Filter the acl:Authorization statements so that they correspond only to statements that apply to
+        // the target (or acl-bearing ancestor) resource path or rdf:type.
+        // Then, assign all acceptable acl:mode values to the relevant acl:agent values: this creates a UNION
+        // of acl:modes for each particular acl:agent.
+        final Map<String, Set<String>> effectiveRoles = new HashMap<>();
+        authorizations.stream()
+            .filter(x -> checkAccessTo.test(x) || checkAccessToClass.test(x))
+            .forEach(x -> {
+                x.getAgents().stream()
+                    .forEach(y -> {
+                        effectiveRoles.putIfAbsent(y, new HashSet<>());
+                        effectiveRoles.get(y).addAll(
+                            x.getModes().stream()
+                                        .map(URI::toString)
+                                        .collect(Collectors.toList()));
+                    });
+            });
+
+        // Transform the effectiveRoles from a Set to a List.
+        return effectiveRoles.entrySet().stream()
+            .map(x -> new AbstractMap.SimpleEntry<>(x.getKey(), new ArrayList<>(x.getValue())))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    /**
+     *  This is a function for generating a Predicate that filters WebACAuthorizations according
+     *  to whether the given acl:accessToClass values contain any of the rdf:type values provided
+     *  when creating the predicate.
+     */
+    private Function<List<String>, Predicate<WebACAuthorization>> accessToClass = uris -> {
+        return x -> {
```

Comment:
@ajs6f, my IDE is telling me there is a better lambda syntax for this logic. Does anything jump out at you?